client/consensus/manual-seal/src/lib.rs

```diff
@@ -0,0 +1,437 @@
+			allow_missing_state: false,
+			import_existing: false,
+		};
+
+		Ok((import_params, None))
+	}
+}
+
+/// Instantiate the import queue for the manual seal consensus engine.
+pub fn import_queue<B: BlockT>(block_import: BoxBlockImport<B>) -> BasicQueue<B>
+{
+	BasicQueue::new(
+		ManualSealVerifier,
+		block_import,
+		None,
+		None,
+	)
+}
+
+/// Creates the background authorship task for the manual seal engine.
+pub async fn run_manual_seal<B, CB, E, A, C, S, H>(
+	mut block_import: BoxBlockImport<B>,
+	mut env: E,
+	back_end: Arc<CB>,
+	basic_pool: Arc<BasicPool<A, B>>,
+	mut seal_block_channel: S,
+	select_chain: C,
+	inherent_data_providers: inherents::InherentDataProviders,
+)
+	where
+		B: BlockT + 'static,
+		H: Hasher<Out=<B as BlockT>::Hash>,
+		CB: ClientBackend<B, H> + 'static,
+		E: Environment<B> + 'static,
+		E::Error: std::fmt::Display,
+		<E::Proposer as Proposer<B>>::Error: std::fmt::Display,
+		A: txpool::ChainApi<Block=B, Hash=<B as BlockT>::Hash> + 'static,
+		S: Stream<Item=EngineCommand<<B as BlockT>::Hash>> + Unpin + 'static,
+		C: SelectChain<B> + 'static,
+{
+	let pool = basic_pool.pool();
+
+	while let Some(command) = seal_block_channel.next().await {
+		match command {
+			EngineCommand::SealNewBlock {
+				create_empty,
+				parent_hash,
+				sender
+			} => {
+				if pool.status().ready == 0 && !create_empty {
+					rpc::send_result(sender, Err(Error::EmptyTransactionPool));
+					continue
+				}
+
+				// get the header to build this new block on.
+				// use the parent_hash supplied via `EngineCommand`
+				// or fetch the best_block.
+				let header = match parent_hash {
+					Some(hash) => {
+						back_end.blockchain().header(BlockId::Hash(hash))
+							.map_err(Error::from)
+							.and_then(|header| {
+								match header {
+									Some(header) => Ok(header),
+									None => Err(Error::BlockNotFound)
+								}
+							})
+					}
+					None => select_chain.best_chain().map_err(Error::from)
+				};
+
+				let header = match header {
+					Err(e) => {
+						rpc::send_result(sender, Err(e));
+						continue
+					}
+					Ok(h) => h,
+				};
+
+				let mut proposer = match env.init(&header) {
+					Err(err) => {
+						rpc::send_result(sender, Err(Error::ProposerError(format!("{}", err))));
+						continue
+					}
+					Ok(p) => p,
+				};
+
+				let id = match inherent_data_providers.create_inherent_data() {
+					Err(err) => {
+						rpc::send_result(sender, Err(err.into()));
+						continue
+					}
+					Ok(id) => id,
+				};
+
+				let result = proposer.propose(
+					id,
+					Default::default(),
+					Duration::from_secs(5),
+				).await;
+
+				let (params, header) = match result {
+					Ok(block) => {
+						let (header, body) = block.deconstruct();
+						(BlockImportParams {
+							origin: BlockOrigin::Own,
+							header: header.clone(),
+							justification: None,
+							post_digests: Vec::new(),
+							body: Some(body),
+							finalized: false,
```

Comment:
I feel like this should be an option from the `SealNewBlock` message. Otherwise, you would always have to send 2 messages to get an instantly-finalized block.